Proofs/Popcount32.lean

```diff
@@ -0,0 +1,129 @@
+   (0x4005e8#64 , 0x0b000020#32), -- add w0, w1, w0
+   (0x4005ec#64 , 0xb9000fe0#32), -- str w0, [sp, #12]
+   (0x4005f0#64 , 0xb9400fe0#32), -- ldr w0, [sp, #12]
+   (0x4005f4#64 , 0x53047c01#32), -- lsr w1, w0, #4
+   (0x4005f8#64 , 0xb9400fe0#32), -- ldr w0, [sp, #12]
+   (0x4005fc#64 , 0x0b000020#32), -- add w0, w1, w0
+   (0x400600#64 , 0x1200cc01#32), -- and w1, w0, #0xf0f0f0f
+   (0x400604#64 , 0x3200c3e0#32), -- mov w0, #0x1010101
+   (0x400608#64 , 0x1b007c20#32), -- mul w0, w1, w0
+   (0x40060c#64 , 0x53187c00#32), -- lsr w0, w0, #24
+   (0x400610#64 , 0xb9000fe0#32), -- str w0, [sp, #12]
+   (0x400614#64 , 0xb9400fe0#32), -- ldr w0, [sp, #12]
+   (0x400618#64 , 0x910043ff#32), -- add sp, sp, #0x10
+   (0x40061c#64 , 0xd65f03c0#32)] -- ret
+
+
+#genStepTheorems popcount32_program namePrefix:="popcount32_" thmType:="fetch"
+
+-- #guard_msgs in
+-- #check popcount32_fetch_0x4005b4
+
+#genStepTheorems popcount32_program namePrefix:="popcount32_" thmType:="decodeExec"
+
+#genStepTheorems popcount32_program namePrefix:="popcount32_" thmType:="step" `state_simp_rules
+
+theorem popcount32_sym_no_error (s0 s_final : ArmState)
+  (h_s0_pc : read_pc s0 = 0x4005b4#64)
+  (h_s0_program : s0.program = popcount32_program)
+  (h_s0_sp_aligned : CheckSPAlignment s0)
+  (h_s0_err : read_err s0 = StateError.None)
+  (h_run : s_final = run 27 s0) :
+  read_err s_final = StateError.None := by
+  -- Prelude
+  simp_all only [state_simp_rules, -h_run]
+  -- Symbolic Simulation
+  sym1_i_n 0 27 h_s0_program
+  try (clear h_step_1 h_step_2 h_step_3 h_step_4;
+       clear h_step_5 h_step_6 h_step_7 h_step_8;
+       clear h_step_9 h_step_10;
+       clear h_step_11 h_step_12 h_step_13 h_step_14;
+       clear h_step_15 h_step_16 h_step_17 h_step_18;
+       clear h_step_19 h_step_20;
+       clear h_step_21 h_step_22 h_step_23 h_step_24;
+       clear h_step_25 h_step_26)
+  -- Final Steps
+  unfold run at h_run
+  subst s_final
+  rw [h_s27_err]
+  done
+
+theorem popcount32_sym_meets_spec (s0 s_final : ArmState)
+  (h_s0_pc : read_pc s0 = 0x4005b4#64)
+  (h_s0_program : s0.program = popcount32_program)
+  (h_s0_sp_aligned : CheckSPAlignment s0)
+  (h_s0_err : read_err s0 = StateError.None)
+  (h_run : s_final = run 27 s0) :
+  read_gpr 32 0#5 s_final = popcount32_spec (read_gpr 32 0#5 s0) ∧
+  read_err s_final = StateError.None := by
+  -- Prelude
+  simp_all only [state_simp_rules, -h_run]
+  -- Symbolic Simulation
+  sym1_i_n 0 27 h_s0_program
+  try (clear h_step_1 h_step_2 h_step_3 h_step_4;
+       clear h_step_5 h_step_6 h_step_7 h_step_8;
+       clear h_step_9 h_step_10;
+       clear h_step_11 h_step_12 h_step_13 h_step_14;
+       clear h_step_15 h_step_16 h_step_17 h_step_18;
+       clear h_step_19 h_step_20;
+       clear h_step_21 h_step_22 h_step_23 h_step_24;
+       clear h_step_25 h_step_26)
+  -- Final Steps
+  unfold run at h_run
+  subst s_final
+  unfold popcount32_spec
+  sorry
```

Comment:
@nwetzler It'd be good to say why this proof is skipped right now. I assume it's because the instruction effects aren't aggregated automatically to obtain the program effects, right?

Reply:
Ahh, sorry, I forgot to save in VSCode before I pushed.  Meant to take this theorem out until it was ready.